cdap-common/src/main/java/co/cask/cdap/common/resource/ResourceBalancerService.java

```diff
@@ -195,8 +195,21 @@ public void onChange(Collection<PartitionReplica> partitionReplicas) {
         @Override
         public void finished(Throwable failureCause) {
           if (service != null) {
-            service.stopAndWait();
-            service = null;
+            try {
+              service.stopAndWait();
+              service = null;
+            } finally {
+              // we need to set completion status for the service, if failure case is non-null set as exception.
+              if (failureCause != null) {
+                LOG.error("No more changes will be notified to the service: {}.", serviceName, failureCause);
+                completion.setException(failureCause);
+              } else {
+                LOG.info("No more changes will be notified to the service: {}.", serviceName);
+                completion.set(null);
+              }
+            }
+          } else {
+            completion.set(null);
```

Comment:
I don't see how the `completion` is being used in the code. Do you know how it is being used? All I see is setting the state of the `completion` but no one is trying to get it.